### PR TITLE
[v5] Mount improvements: automatic parameter inference/checking

### PR DESCRIPTION
**Pull request by keichinger**
| Q             | A
| ------------- | --------------------------------------------------------------------- |
| Bug fix?      | no                                                                |
| New feature?  | yes                   |
| Improvement?  |no     |
| BC breaks?    | yes                                                               |
| Deprecations? |no     |
| Docs PR       | **missing**                                   |

This PR splits the `mount` method up (again) into three distinct methods:
- `mount` for mounting `function`s
- `mountClass` for mounting `StandaloneComponent`s
- `mountJsx` for mounting Preact components (functional and class)

This change has sadly been necessary, as TypeScripts generic type inference couldn't support a new feature: automatic parameter inference/checking for the component you were trying to mount. In case there is no additional component parameters, the corresponding `mojave.*MountOptions::params` type will be `never`, to enforce this as well.

There is one caveat to this feature, which I'm not yet sure how to prevent/implement it right now, which is TypeScript isn't throwing any errors when I don't define the `options.params` property. However, this is a known limitation (for now) which I greatly accept as the new feature is better for overall DX and reduces the amount of accidental errors. Other than that, we're currently not throwing any errors as well, so this is not even a regression and thus should be fine.

Here are a few examples of what's working and what is (currently) still not yet possible to support:

## `mount` with `function`s

Example component

```typescript
function someMountableFunction(el: HTMLElement, a: boolean): void
{
}
```

`mount` now is aware of the second, boolean parameter and correctly throws errors when too few or too many parameters have been passed. It extracts any additional `function` from the passed-in function and uses these as type for `mojave.FunctionMountOptions::params`:

```typescript
mount(".foo", someMountableFunction);                           // no errors -> false positive & known limitation!
mount(".foo", someMountableFunction, { type: "func" });         // no errors -> false positive & known limitation!
mount(".foo", someMountableFunction, { type: "func", params: [  // no errors -> correct!
    true,
]});
mount(".foo", someMountableFunction, { type: "func", params: [  // errors, too many parameters -> correct!
    true,
    "bla", // unknown additional parameter
    42,    // unknown additional parameter
]});
```

## `mountClass` with `StandaloneComponent`

Example component

```typescript
class SomeMountableComponent extends StandaloneComponent
{
    private el: HTMLElement;
    private a: boolean;

    public constructor(el: HTMLElement, a: boolean)
	{
        this.el = el;
        this.a = a;
    }

    /**
     * @inheritDoc
     */
    public init(): void
    {
    }
}
```

Here, mounting `SomeMountableComponent` using `mountClass` works the same way as `mount` does: It automatically extracts additional constructor parameters from the passed-in `StandaloneComponent`, which is then used as type for `mojave.ClassMountOptions::params` to check against.

```typescript
mountClass(".foo", SomeMountableComponent);                             // no errors -> false positive & known limitation!
mountClass(".foo", SomeMountableComponent, { type: "class" });          // no errors -> false positive & known limitation!
mountClass(".foo", SomeMountableComponent, { type: "class", params: [   // no errors -> correct!
    true,
]});
mountClass(".foo", SomeMountableComponent, { type: "class", params: [   // errors, too many parameters -> correct!
    true,
    "bla", // unknown additional parameter
    42,    // unknown additional parameter
]});
```

## `mountJsx` with Preact components

Last but not least:

```typescript
//
// Functional component
//
type PreactFunctionalProps = {
    a: string;
    b: number;
}

function SomePreactFunctionalComponent (props: PreactFunctionalProps) : JSX.Element
{
    return null;
}

//
// Class component
//
type SomePreactClassComponentProps = {
    a: string;
    b: number;
}

class SomePreactClassComponent extends Component<SomePreactClassComponentProps, {}>
{
    /**
     * @inheritDoc
     */
    public render(props: Readonly<SomePreactClassComponentProps>, state: Readonly<{}>): preact.ComponentChild
    {
        return null;
    }
}
```

The story here is very similar. The only difference, as previously, is that `mojave.ComponentMountOptions::params`'s type is now the corresponding props-type of the given component. So in the case of `SomePreactFunctionalComponent` the type is `PreactFunctionalProps`, for `SomePreactClassComponent` it's `SomePreactClassComponentProps`.

```typescript
//
// Functional component
//
mountJsx(".foo", SomePreactFunctionalComponent);                            // no errors -> false positive & known limitation!
mountJsx(".foo", SomePreactFunctionalComponent, { type: "jsx" });           // no errors -> false positive & known limitation!
mountJsx(".foo", SomePreactFunctionalComponent, { type: "jsx", params: {    // no errors -> correct!
    a: "foo",
    b: 42,
}});
mountJsx(".foo", SomePreactFunctionalComponent, { type: "jsx", params: {    // errors, too many parameters -> correct!
    a: "foo",
    b: 42,
    asdf: "bla", // unknown additional parameter
}});

//
// Class component
//
mountJsx(".foo", SomePreactClassComponent);                                 // no errors -> false positive & known limitation!
mountJsx(".foo", SomePreactClassComponent, { type: "jsx" });                // no errors -> false positive & known limitation!
mountJsx(".foo", SomePreactClassComponent, { type: "jsx", params: {         // no errors -> correct!
    a: "foo",
    b: 42,
}});
mountJsx(".foo", SomePreactClassComponent, { type: "jsx", params: {         // errors, too many parameters -> correct!
    a: "foo",
    b: 42,
    asdf: "bla", // unknown additional parameter
}});
```

### CODE REVIEW

File: @types/mojave.d.ts [hunk 1]
```diff
@@ -8,7 +9,7 @@ declare namespace mojave
         }
     };
     export type MountableType = "func" | "jsx" | "class";
-    export type MountableFunction = (element: HTMLElement, ...args: any[]) => void;
+    export type MountableFunction = (element: HTMLElement, ...args: any[]) => any;
```
Comment (apfelbox) [inline @types/mojave.d.ts]: Why is that change necessary?

File: UPGRADE.md [hunk 2]
```diff
@@ -1,3 +1,17 @@
...
+==========
+
+* `mount` has been split into three methods:
+    * `mount` for mounting functions
+    * `mountClass` for mounting `StandaloneComponent`s
+    * `mountJsx` for mounting Preact components (functional and class)
+* Corresponding to the changes in `mount`, we had to change the mounting option types to add support for automatic parameter inference:
+    * `mojave.ClassMountOptions` has been made generic: `mojave.ClassMountOptions<T extends mojave.MountableClass>`
+    * `mojave.FunctionMountOptions` has been made generic: `mojave.FunctionMountOptions<T extends mojave.MountableFunction>`
+    * `mojave.ComponentMountOptions` has been made generic: `mojave.ComponentMountOptions<T extends ComponentClass<any> | FunctionComponent<any>>`
```
Comment (apfelbox) [inline UPGRADE.md]: You should not copy the changelog here (as that's what the changelog is for), but instead describe on how to upgrade (that's why this file is called "upgrade").

File: mount/index.ts [hunk 3]
```diff
@@ -84,18 +100,16 @@ function doMount (elements: HTMLElement[], mountable: mojave.Mountable, rawOptio
             }
             else if ("class" === options.type)
             {
-                let standaloneOptions = options as mojave.ClassMountOptions;
-                const mounted = new mountableAny(node, ...(standaloneOptions.params || []));
+                let mountableClass = mountable as mojave.MountableClass;
+                let standaloneOptions = options as mojave.ClassMountOptions<any>;
+                const mounted = new mountableClass(node, ...(standaloneOptions.params || []));
```
Comment (apfelbox) [inline mount/index.ts]: Do we really need this change? Isn't it enough to change the type declaration at the "interface functions" (= exported functions)

File: @types/mojave.d.ts (hunk 1 above)
Comment (keichinger) [reply @types/mojave.d.ts]: It seemed unnecessary restrictive that we enforce the function to return `void`. In theory they could return whatever they want, it's just that `mount*` doesn't handle/return it.

File: mount/index.ts (hunk 3 above)
Comment (apfelbox) [reply mount/index.ts]: Ah, you just moved the variable into this `if`. Nevermind then

**Comment by apfelbox** [comment]
With this change we must get rid of the `type` parameter and ideally with the `if`-parts in `doMount` as well.

Also please target `master` and branch of a `master-v4`. `master` is always the latest version.

**Comment by keichinger** [comment]
Apparently GitHub has automatically closed my PR because I've renamed the `master` branches. Well, I've opened a new PR #226 with the requested changes.
